core/src/main/java/org/elasticsearch/node/Node.java

```diff
@@ -523,15 +523,26 @@ static void checkForIndexDataInDefaultPathData(
 
         boolean clean = true;
         for (final String defaultPathData : Environment.DEFAULT_PATH_DATA_SETTING.get(settings)) {
-            final Path nodeDirectory = NodeEnvironment.resolveNodePath(getPath(defaultPathData), nodeEnv.getNodeLockId());
-            if (Files.exists(nodeDirectory) == false) {
+            final Path defaultNodeDirectory = NodeEnvironment.resolveNodePath(getPath(defaultPathData), nodeEnv.getNodeLockId());
+            if (Files.exists(defaultNodeDirectory) == false) {
                 continue;
             }
-            final NodeEnvironment.NodePath nodePath = new NodeEnvironment.NodePath(nodeDirectory);
+
+            boolean includedInPathData = false;
+            for (final NodeEnvironment.NodePath dataPath : nodeEnv.nodePaths()) {
+                includedInPathData |= Files.isSameFile(dataPath.path, defaultNodeDirectory);
```

Comment:
I pushed 7d5331765f0a4e2dbf035b53aff414428ed5b363.

It is incredibly fortunate that you asked for this. As a result of asking for this, I made the necessary change and tried to run `gradle :core:test -Dtests.class=org.elasticsearch.node.NodeTests -Dtests.iters=128` from the command-line to ensure that the change did not break anything. This failed, Gradle told me that no tests were executed. It turns out that `org.elasticsearch.node.NodeTests` was sitting in the test framework in src/main/java, so these tests were not executing. It appears the class has been like this since July of last year. Fortunately, nothing is broken. I moved this test class to the appropriate location in core.